fix: clone repository before getting working dir

ghaiszaher opened the pull request:
## what

* Fixes https://github.com/runatlantis/atlantis/issues/3855
* As described in the issue, when running `atlantis plan` on specific projects, the repository is not cloned first
* This is reproducible when the first command you run in a PR is `atlantis plan -d <dir>` or `atlantis plan -p <project>` in either one of the following situations:
  * `ATLANTIS_DISABLE_AUTOPLAN = true` is used
  * `ATLANTIS_DISABLE_AUTOPLAN_LABEL = "no-autoplan"` is used and the pr is labeled with `no-autoplan`
  * A draft PR
  * A PR that was opened while atlantis/webhook is not working
* Solution: clone the PR earlier

## why
* It is necessary to make sure the plan command clones the repo at all times and shouldn't expect it to be already cloned (e.g. if the local directory had to be cleared for some reason)

## tests
I have done the tests by running the service locally and using a webhook in a private GitHub project:
* `sudo rm -r .atlantis`
* Restart atlantis
* Comment `atlantis plan -d my-dir`

| Before | After |
|--------|--------|
|<img width="920" alt="image" src="https://github.com/runatlantis/atlantis/assets/20379416/5fb7306e-1dbd-47e0-aec3-8c041cd2ba46"> | <img width="922" alt="image" src="https://github.com/runatlantis/atlantis/assets/20379416/a79d2f75-1d4c-47e0-b2dc-c42bfe4bc23e"> |

## references
* https://github.com/runatlantis/atlantis/issues/3855

ghaiszaher commented:
> I don't know all the inner workings of Atlantis very well, but the comment above the lock for `buildAllCommandsByCfg` suggests that the workspace should be locked before cloning?
> 
> https://github.com/runatlantis/atlantis/blob/793aafb2358792f68cb507d0f18baada4c3e993f/server/events/project_command_builder.go#L371-L375
> 
> Is there any downside to moving the TryLock up as well?

I missed this, thank you. I don't see any downsides: ab78b2b. I'll test it again and report if I see any issues.

ghaiszaher commented:
> I'll test it again and report if I see any issues.

No issues

nitrocode commented:
Nice work fixing this. Thank you for contributing the pr.

Would it be possible to add a golang test for this as well to ensure this feature doesn't break in the future?

ghaiszaher commented on server/events/project_command_builder.go:
```diff
@@ -509,6 +509,19 @@ func (p *DefaultProjectCommandBuilder) buildProjectPlanCommand(ctx *command.Cont
+	defer unlockFn()
+
+	ctx.Log.Debug("cloning repository")
+	_, _, err = p.WorkingDir.Clone(ctx.HeadRepo, ctx.Pull, DefaultWorkspace)
```
It's necessary to clone `DefaultWorkspace` at this point and not `workspace`

ghaiszaher commented on server/events/project_command_builder.go:
```diff
@@ -580,17 +593,12 @@ func (p *DefaultProjectCommandBuilder) buildProjectPlanCommand(ctx *command.Cont
-	_, _, err = p.WorkingDir.Clone(ctx.HeadRepo, ctx.Pull, workspace)
-	if err != nil {
-		return pcc, err
+	if DefaultWorkspace != workspace {
+		ctx.Log.Debug("cloning repository with workspace %s", workspace)
+		_, _, err = p.WorkingDir.Clone(ctx.HeadRepo, ctx.Pull, workspace)
+		if err != nil {
+			return pcc, err
+		}
```
This whole block is unnecessary because this call happens later anyway, but I kept it just in case it's needed for cases that I didn't notice

ghaiszaher commented:
> Nice work fixing this. Thank you for contributing the pr.
> 
> Would it be possible to add a golang test for this as well to ensure this feature doesn't break in the future?

@nitrocode I adjusted the e2e tests here: 9f4d3e1
I added 2 test cases with disabled auto-plan and with no pre-workflow webhooks (because pre-workflow webhooks are already cloning the default workspace).

Please also check my comments in [server/events/project_command_builder.go](https://github.com/runatlantis/atlantis/pull/3867/files#diff-d76e071cce4d7fd5d1c03f8220cc2ae2380e6e0121bb907f2f3a3bf82ef5b8d4)

nitrocode reviewed (approved):
Thanks @ghaiszaher . Approved.

Ill defer to a follow up review before merging.
